### PR TITLE
Accept branches as requirement in metadata.yml

### PR DESCRIPTION
**Pull request by bennythejudge**
Allows to specify a branch as a requirement in metadata.yml.

Example:

```
formula: ministryofjustice/test-formula

dependencies:
- ministryofjustice/php-fpm-formula==FixFpmOnReboot
```

### CODE REVIEW

**Comment by niallcreech** [comment]
Looks good, couple of things, is there a reason why 
obj = {
+                    'commit': {
+                        'url': branch_data['commit']['sha'],
+                        'sha': branch_data['commit']['sha'],
+                    },
+                    'name': branch_data['name'],
+                }

rather than just obj = branch_data
?

File: shaker/libs/github.py [hunk 1]
```diff
@@ -363,17 +416,62 @@ def resolve_constraint_to_object(org_name, formula_name, constraint):
...
+        # is it a branch (i.e. not a version)
+        if not parsed_constraint['version']:
+            branch_name = parsed_constraint['tag']
+            shaker.libs.logger.Logger().debug("github::resolve_constraint_to_object: "
+                                              "There is no version, assuming this is "
+                                              "a branch, name: '%s'"
+                                              % (branch_name))
+            branch_data = get_branch_data(org_name, formula_name, branch_name)
+            if branch_data:
+                obj = {
```
Comment (niallcreech) [inline shaker/libs/github.py]: Any reason not to just obj = branch_data

File: tests/libs/test_github.py [hunk 2]
```diff
@@ -33,6 +33,84 @@ class TestGithub(unittest.TestCase):
         },
     ]
 
+    _sample_response_branches = {
```
Comment (niallcreech) [inline tests/libs/test_github.py]: Maybe best to make this as small as we need, a couple of entries with name commit and sha should probably do

**Comment by niallcreech** [comment]
Also a couple of comments on reducing the mock data we're using, keeping them as minimal as possible

File: shaker/libs/github.py [hunk 3]
```diff
@@ -94,7 +94,8 @@ def parse_semver_tag(tag):
                 "patch": patch_version,
                 "postfix": (posfix-tag)
             }
-            Return an empty list if the tag could not be parsed.
+            Return a dictionary with all fields set to None if
+            the tag could not be parsed.
```
Comment (ashb) [inline shaker/libs/github.py]: It looks like this function is still returning a list, not a dictionary.

File: tests/libs/test_github.py (hunk 2 above)
Comment (bennythejudge) [inline tests/libs/test_github.py]: I have edited the mockup json to contain the minimum required entries.

File: shaker/libs/github.py (hunk 1 above)
Comment (bennythejudge) [inline shaker/libs/github.py]: When requesting branch details, Github returns a lengthy json which contains a lot of entries which are not used in this context. 
I decided to get the essential values, to keep the object similar to the one returned when the requirement is a version.

**Pull request merged**